spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala

```diff
@@ -505,6 +506,84 @@ trait DeltaTypeWideningNestedFieldsTests {
   }
 }
 
+/** Tests covering type widening compatibility with other delta features. */
+trait DeltaTypeWideningCompatibilityTests {
+  self: DeltaTypeWideningTestMixin with QueryTest with DeltaDMLTestUtils =>
+
+  import testImplicits._
+
+  test("reading CDF with a type change") {
+    withSQLConf((DeltaConfigs.CHANGE_DATA_FEED.defaultTablePropertyKey, "true")) {
+      sql(s"CREATE TABLE delta.`$tempPath` (a smallint) USING DELTA")
+    }
+    append(Seq(1, 2).toDF("a").select($"a".cast(ShortType)))
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN a TYPE int")
+    append(Seq(3, 4).toDF("a"))
+
+    def readCDF(start: Long, end: Long): DataFrame =
+      CDCReader
+        .changesToBatchDF(deltaLog, start, end, spark)
+        .drop(CDCReader.CDC_COMMIT_TIMESTAMP)
+        .drop(CDCReader.CDC_COMMIT_VERSION)
+
+    checkErrorMatchPVals(
+      exception = intercept[DeltaUnsupportedOperationException] {
+        readCDF(start = 1, end = 1).collect()
+      },
+      errorClass = "DELTA_CHANGE_DATA_FEED_INCOMPATIBLE_DATA_SCHEMA",
+      parameters = Map(
+        "start" -> "1",
+        "end" -> "1",
+        "readSchema" -> ".*",
+        "readVersion" -> "3",
+        "incompatibleVersion" -> "1",
+        "config" -> ".*defaultSchemaModeForColumnMappingTable"
+      )
+    )
+    checkAnswer(readCDF(start = 3, end = 3), Seq(Row(3, "insert"), Row(4, "insert")))
+  }
+
+  test("change column type and rename it with column mapping") {
+    withSQLConf((DeltaConfigs.COLUMN_MAPPING_MODE.defaultTablePropertyKey, IdMapping.name)) {
+      sql(s"CREATE TABLE delta.`$tempPath` (a byte) USING DELTA")
+    }
+    // Add some data and change type of column `a`.
+    append(Seq(1).toDF("a").select($"a".cast(ByteType)))
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN a TYPE smallint")
+    append(Seq(2).toDF("a").select($"a".cast(ShortType)))
+    assert(readDeltaTable(tempPath).schema("a").dataType === ShortType)
+    checkAnswer(sql(s"SELECT a FROM delta.`$tempPath`"), Seq(Row(1), Row(2)))
+
+    // Rename column `a` to `b`, add more data.
+    sql(s"ALTER TABLE delta.`$tempPath` RENAME COLUMN a TO b")
+    assert(readDeltaTable(tempPath).schema("b").dataType === ShortType)
+    checkAnswer(sql(s"SELECT b FROM delta.`$tempPath`"), Seq(Row(1), Row(2)))
+    append(Seq(3).toDF("b").select($"b".cast(ShortType)))
+    checkAnswer(sql(s"SELECT b FROM delta.`$tempPath`"), Seq(Row(1), Row(2), Row(3)))
+
+    // Change column type again, add more data.
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN b TYPE int")
+    assert(readDeltaTable(tempPath).schema("b").dataType === IntegerType)
+    append(Seq(4).toDF("b"))
+    checkAnswer(sql(s"SELECT b FROM delta.`$tempPath`"), Seq(Row(1), Row(2), Row(3), Row(4)))
```

Comment:
Do we also have a test for dropping the feature with a column that was renamed after its type was changed?